src/devices/cpu/tms34010/34010ops.hxx

```diff
@@ -2392,10 +2392,51 @@ void tms340x0_device::rpix_b(uint16_t op)
 	RPIX(B);
 }
 
+constexpr std::uint32_t log2(std::uint32_t n)
+{
+	return (n > 1) ? 1 + log2(n >> 1) : 0;
+}
+
 void tms340x0_device::setcdp(uint16_t op)
 {
 	if (!m_is_34020) { unimpl(op); return; }
-	logerror("020:setcdp\n");
+	off_t dptch = DPTCH();
+
+	// logerror("020:setcdp: PC=%x, DPTCH()=%x\n", m_pc, dptch);
+
+	// Check whether we're dealing with an even number
+	if ((dptch & 1) == 0) {
+		switch(__builtin_popcount(dptch))
+		{
+			// .. only single bit set, pitch is power of two!
+			case 1:
+			{
+				uint32_t val = log2(dptch);
+				m_convdp = ~val & 0x1f;
+				COUNT_CYCLES(4);
+				return;
+			}
+			// .. two bits, we can decompose it to sum of two power of two numbers
+			case 2:
+			{
+				uint8_t b1 = __builtin_ffs(dptch) - 1;
+				uint8_t b2 = __builtin_ffs(dptch & ~(1 << b1)) - 1;
+
+				uint32_t val1 = log2(1 << b1);
```

Comment:
The base 2 logarithm of `1 << x` is `x`.  Think about it: `1 << x` is the same operation as two to the power of `x` (let's call it `2 ** x`), so `log2(1 << x)` is equivalent to `log2(2 ** x)`, which is equal to `x`.  You definitely need to re-think your maths here.

Reply:
Thanks, I had misread the manual. Testcase with 4 sample from tms34020 manual is here:

https://gist.github.com/plaes/dec4b76ffc4b720726f21031d8d399ac